driver/src/driver/scheduler/evm.rs

```diff
@@ -0,0 +1,279 @@
+//! Implementation of an EVM-based scheduler that retrieves current batch and
+//! batch duration information directly from the EVM instead of system time.
+
+use super::{AuctionTimingConfiguration, Scheduler, BATCH_DURATION};
+use crate::contracts::stablex_contract::StableXContract;
+use crate::driver::stablex_driver::{DriverResult, StableXDriver};
+use anyhow::Result;
+use log::{debug, error, info, warn};
+use std::thread;
+use std::time::Duration;
+
+/// The amount of time the scheduler should wait between polling.
+const POLL_TIMEOUT: Duration = Duration::from_secs(5);
+
+/// An EVM-based scheduler for the exchange driver.
+pub struct EvmScheduler<'a> {
+    exchange: &'a dyn StableXContract,
+    driver: &'a dyn StableXDriver,
+    config: AuctionTimingConfiguration,
+    last_batch: Option<u32>,
+}
+
+impl<'a> EvmScheduler<'a> {
+    /// Creates a new EVM-based scheduler from a configuration.
+    pub fn new(
+        exchange: &'a dyn StableXContract,
+        driver: &'a dyn StableXDriver,
+        config: AuctionTimingConfiguration,
+    ) -> Self {
+        EvmScheduler {
+            driver,
+            exchange,
+            config,
+            last_batch: None,
+        }
+    }
+
+    /// Creates a new scheduler with the default configuration.
+    #[cfg(test)]
+    pub fn with_defaults(exchange: &'a dyn StableXContract, driver: &'a dyn StableXDriver) -> Self {
+        EvmScheduler::new(exchange, driver, AuctionTimingConfiguration::default())
+    }
+
+    /// Runs the scheduler for a single iteration.
+    fn step(&mut self) -> Result<()> {
+        let batch_id = self.current_solving_batch()?;
+        if self
+            .last_batch
+            .map(|last_batch| batch_id <= last_batch)
+            .unwrap_or_default()
+        {
+            debug!("skipping already processed batch {}", batch_id);
+            return Ok(());
+        }
+
+        let time_remaining = self.exchange.get_current_auction_remaining_time()?;
+        // NOTE: We need to take into account the asynchronous nature of web3
```

Comment:
Fine for now. We could deploy a wrapper contract that gives us this information? Or we could incorporate the batching API into ethcontracts which would allow to run multiple "calls" in the same context.

Reply:
If its OK with you, I'll do that as a separate PR.

Reply:
@fleupold Would this be a view method that I could add to `dex-contracts`?

Reply:
Yes, I think it could be added to https://github.com/gnosis/dex-contracts/blob/master/contracts/BatchExchangeViewer.sol